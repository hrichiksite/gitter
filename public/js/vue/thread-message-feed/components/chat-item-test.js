const mount = require('../../__test__/vuex-mount');
const momentTimezone = require('moment-timezone');
const {
  createSerializedMessageFixture,
  createSerializedRoomFixture
} = require('../../__test__/fixture-helpers');
const { default: ChatItem } = require('./chat-item.vue');

jest.mock('../../../components/unread-items-client');
jest.mock('../../../utils/is-mobile');
const unreadItemsClient = require('../../../components/unread-items-client');

describe('thread-message-feed chat-item', () => {
  momentTimezone.tz.setDefault('Europe/London');
  const message = createSerializedMessageFixture();
  const defaultProps = {
    message,
    useCompactStyles: false
  };
  const addRoomToStore = store => {
    const room = createSerializedRoomFixture('abc/def');
    store.state.roomMap = { [room.id]: room };
    store.state.displayedRoomId = room.id;
  };

  describe('snapshot', () => {
    it('with default props', () => {
      const { wrapper } = mount(ChatItem, defaultProps, addRoomToStore);
      expect(wrapper.element).toMatchSnapshot();
    });
    describe('component flags', () => {
      ['useCompactStyles', 'showItemActions'].forEach(flag => {
        it(flag, () => {
          const { wrapper } = mount(
            ChatItem,
            {
              ...defaultProps,
              [flag]: true
            },
            addRoomToStore
          );
          expect(wrapper.element).toMatchSnapshot();
        });
      });
    });
    describe('message flags', () => {
      ['error', 'loading', 'unread'].forEach(flag => {
        it(flag, () => {
          const { wrapper } = mount(
            ChatItem,
            {
              ...defaultProps,
              message: { ...message, [flag]: true }
            },
            addRoomToStore
          );
          expect(wrapper.element).toMatchSnapshot();
        });
      });
    });
    it('highlighted - scrolls into view', () => {
      const scrollIntoViewMock = jest.fn();
      const { wrapper } = mount(
        ChatItem,
        {
          ...defaultProps,
          message: { ...message, highlighted: true }
        },
        addRoomToStore,
        { methods: { scrollIntoView: scrollIntoViewMock } }
      );
      expect(wrapper.element).toMatchSnapshot();
      expect(scrollIntoViewMock.mock.calls[0]).toEqual(['smooth', 'center']);
    });
    it('deleted (empty) message', () => {
      const { wrapper } = mount(
        ChatItem,
        {
          ...defaultProps,
          message: { ...message, text: '', html: '' }
        },
        addRoomToStore
      );
      expect(wrapper.element).toMatchSnapshot();
    });
    it('not logged in user (no message actions)', () => {
      const { wrapper } = mount(
        ChatItem,
        {
          ...defaultProps,
          message: { ...message, text: '', html: '' }
        },
        store => {
          addRoomToStore(store);
          store.state.isLoggedIn = false;
        }
      );
      expect(wrapper.element).toMatchSnapshot();
    });
  });
  describe('editing', () => {
    let wrapper, stubbedActions;
    beforeEach(() => {
      const mountResult = mount(
        ChatItem,
        {
          ...defaultProps,
          message: { ...message, text: '', html: '' }
        },
        store => {
          addRoomToStore(store);
          store.state.threadMessageFeed = {
            messageEditState: {
              id: message.id,
              text: 'updated text'
            }
          };
        }
      );
      wrapper = mountResult.wrapper;
      stubbedActions = mountResult.stubbedActions;
    });

    it('matches snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('pressing enter submits the message for update', () => {
      wrapper.find({ ref: 'chatItemEditTextArea' }).trigger('keydown.enter');
      expect(stubbedActions.threadMessageFeed.updateMessage).toHaveBeenCalled();
    });
    it('pressing esc cancels editing', () => {
      wrapper.find({ ref: 'chatItemEditTextArea' }).trigger('keydown.esc');
      expect(stubbedActions.threadMessageFeed.cancelEdit).toHaveBeenCalled();
    });
  });
  it('focusedAt - scrolls into view', () => {
    const scrollIntoViewMock = jest.fn();
    mount(
      ChatItem,
      {
        ...defaultProps,
        message: { ...message, focusedAt: { block: 'start', timestamp: Date.now() } }
      },
      addRoomToStore,
      { methods: { scrollIntoView: scrollIntoViewMock } }
    );
    expect(scrollIntoViewMock.mock.calls[0]).toEqual(['auto', 'start']);
  });
  it('unread message reports itself as read when entering viewport', () => {
    const { wrapper } = mount(
      ChatItem,
      {
        ...defaultProps,
        message: { ...message, unread: true }
      },
      addRoomToStore
    );
    wrapper.vm.onViewportEnter();
    expect(unreadItemsClient.markItemRead.mock.calls[0]).toEqual([message.id]);
  });

  it('double tap triggers edit message action on mobile', () => {
    const { wrapper, stubbedActions } = mount(ChatItem, defaultProps, addRoomToStore);
    const isMobile = require('../../../utils/is-mobile');
    isMobile.mockImplementation(() => true);
    // avoiding calling the real action so we don't have to set up the store
    stubbedActions.threadMessageFeed.editMessage.mockImplementation(() => {});
    wrapper.find('.chat-item').trigger('touchend');
    wrapper.find('.chat-item').trigger('touchend');
    expect(stubbedActions.threadMessageFeed.editMessage).toHaveBeenCalledWith(
      expect.anything(),
      message,
      undefined
    );
  });
});
