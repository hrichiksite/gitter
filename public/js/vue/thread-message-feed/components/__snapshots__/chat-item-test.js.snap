// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`thread-message-feed chat-item editing matches snapshot 1`] = `
<intersect-stub>
  <div
    class="chat-item burstStart deleted"
    role="listitem"
  >
    <div
      class="chat-item__container"
    >
      <div
        class="chat-item__aside"
      >
        <div
          class="chat-item__avatar"
        >
          <avatar-stub
            size="s"
            user="[object Object]"
          />
        </div>
      </div>
       
      <chat-item-actions-stub
        message="[object Object]"
      />
       
      <div
        class="chat-item__content"
      >
        <div
          class="chat-item__details"
        >
          <div
            class="chat-item__from"
          >
            Tomas Vik
          </div>
           
          <div
            class="chat-item__username"
          >
            @viktomas_gitlab
          </div>
           
          <a
            class="chat-item__time"
            href="/abc/def?at=5d147ea84dad9dfbc522317a"
            title="May 18, 2016 3:48 AM"
          >
            May 18 2016 03:48
          </a>
           
          <!---->
        </div>
         
        <textarea
          class="trpChatInput"
        />
      </div>
    </div>
  </div>
</intersect-stub>
`;

exports[`thread-message-feed chat-item snapshot component flags showItemActions 1`] = `
<intersect-stub
  showitemactions="true"
>
  <div
    class="chat-item burstStart"
    role="listitem"
  >
    <div
      class="chat-item__container"
    >
      <div
        class="chat-item__aside"
      >
        <div
          class="chat-item__avatar"
        >
          <avatar-stub
            size="s"
            user="[object Object]"
          />
        </div>
      </div>
       
      <chat-item-actions-stub
        message="[object Object]"
      />
       
      <div
        class="chat-item__content"
      >
        <div
          class="chat-item__details"
        >
          <div
            class="chat-item__from"
          >
            Tomas Vik
          </div>
           
          <div
            class="chat-item__username"
          >
            @viktomas_gitlab
          </div>
           
          <a
            class="chat-item__time"
            href="/abc/def?at=5d147ea84dad9dfbc522317a"
            title="May 18, 2016 3:48 AM"
          >
            May 18 2016 03:48
          </a>
           
          <!---->
        </div>
         
        <div
          class="chat-item__text"
        >
          <span
            class="mention"
            data-link-type="mention"
            data-screen-name="MadLittleMods"
          >
            @MadLittleMods
          </span>
            Example message using a bit of  
          <code>
            code
          </code>
           and 
          <strong>
            bold
          </strong>
           to show how 
          <em>
            markdown
          </em>
           is stored.
        </div>
      </div>
    </div>
  </div>
</intersect-stub>
`;

exports[`thread-message-feed chat-item snapshot component flags useCompactStyles 1`] = `
<intersect-stub>
  <div
    class="chat-item burstStart compact"
    role="listitem"
  >
    <div
      class="chat-item__container"
    >
      <div
        class="chat-item__aside"
      >
        <div
          class="chat-item__avatar"
        >
          <avatar-stub
            size="s"
            user="[object Object]"
          />
        </div>
      </div>
       
      <chat-item-actions-stub
        message="[object Object]"
        usecompactstyles="true"
      />
       
      <div
        class="chat-item__content"
      >
        <div
          class="chat-item__details"
        >
          <div
            class="chat-item__from"
          >
            Tomas Vik
          </div>
           
          <div
            class="chat-item__username"
          >
            @viktomas_gitlab
          </div>
           
          <a
            class="chat-item__time"
            href="/abc/def?at=5d147ea84dad9dfbc522317a"
            title="May 18, 2016 3:48 AM"
          >
            May 18 2016 03:48
          </a>
           
          <!---->
        </div>
         
        <div
          class="chat-item__text"
        >
          <span
            class="mention"
            data-link-type="mention"
            data-screen-name="MadLittleMods"
          >
            @MadLittleMods
          </span>
            Example message using a bit of  
          <code>
            code
          </code>
           and 
          <strong>
            bold
          </strong>
           to show how 
          <em>
            markdown
          </em>
           is stored.
        </div>
      </div>
    </div>
  </div>
</intersect-stub>
`;

exports[`thread-message-feed chat-item snapshot deleted (empty) message 1`] = `
<intersect-stub>
  <div
    class="chat-item burstStart deleted"
    role="listitem"
  >
    <div
      class="chat-item__container"
    >
      <div
        class="chat-item__aside"
      >
        <div
          class="chat-item__avatar"
        >
          <avatar-stub
            size="s"
            user="[object Object]"
          />
        </div>
      </div>
       
      <chat-item-actions-stub
        message="[object Object]"
      />
       
      <div
        class="chat-item__content"
      >
        <div
          class="chat-item__details"
        >
          <div
            class="chat-item__from"
          >
            Tomas Vik
          </div>
           
          <div
            class="chat-item__username"
          >
            @viktomas_gitlab
          </div>
           
          <a
            class="chat-item__time"
            href="/abc/def?at=5d147ea84dad9dfbc522317a"
            title="May 18, 2016 3:48 AM"
          >
            May 18 2016 03:48
          </a>
           
          <!---->
        </div>
         
        <div
          class="chat-item__text"
        >
          
          
          
          <i>
            This message was deleted
          </i>
        </div>
      </div>
    </div>
  </div>
</intersect-stub>
`;

exports[`thread-message-feed chat-item snapshot highlighted - scrolls into view 1`] = `
<intersect-stub>
  <div
    class="chat-item burstStart chat-item__highlighted"
    role="listitem"
  >
    <div
      class="chat-item__container"
    >
      <div
        class="chat-item__aside"
      >
        <div
          class="chat-item__avatar"
        >
          <avatar-stub
            size="s"
            user="[object Object]"
          />
        </div>
      </div>
       
      <chat-item-actions-stub
        message="[object Object]"
      />
       
      <div
        class="chat-item__content"
      >
        <div
          class="chat-item__details"
        >
          <div
            class="chat-item__from"
          >
            Tomas Vik
          </div>
           
          <div
            class="chat-item__username"
          >
            @viktomas_gitlab
          </div>
           
          <a
            class="chat-item__time"
            href="/abc/def?at=5d147ea84dad9dfbc522317a"
            title="May 18, 2016 3:48 AM"
          >
            May 18 2016 03:48
          </a>
           
          <!---->
        </div>
         
        <div
          class="chat-item__text"
        >
          <span
            class="mention"
            data-link-type="mention"
            data-screen-name="MadLittleMods"
          >
            @MadLittleMods
          </span>
            Example message using a bit of  
          <code>
            code
          </code>
           and 
          <strong>
            bold
          </strong>
           to show how 
          <em>
            markdown
          </em>
           is stored.
        </div>
      </div>
    </div>
  </div>
</intersect-stub>
`;

exports[`thread-message-feed chat-item snapshot message flags error 1`] = `
<intersect-stub>
  <div
    class="chat-item burstStart syncerror"
    role="listitem"
  >
    <div
      class="chat-item__container"
    >
      <div
        class="chat-item__aside"
      >
        <div
          class="chat-item__avatar"
        >
          <avatar-stub
            size="s"
            user="[object Object]"
          />
        </div>
      </div>
       
      <chat-item-actions-stub
        message="[object Object]"
      />
       
      <div
        class="chat-item__content"
      >
        <div
          class="chat-item__details"
        >
          <div
            class="chat-item__from"
          >
            Tomas Vik
          </div>
           
          <div
            class="chat-item__username"
          >
            @viktomas_gitlab
          </div>
           
          <a
            class="chat-item__time"
            href="/abc/def?at=5d147ea84dad9dfbc522317a"
            title="May 18, 2016 3:48 AM"
          >
            May 18 2016 03:48
          </a>
           
          <!---->
        </div>
         
        <div
          class="chat-item__text"
        >
          <span
            class="mention"
            data-link-type="mention"
            data-screen-name="MadLittleMods"
          >
            @MadLittleMods
          </span>
            Example message using a bit of  
          <code>
            code
          </code>
           and 
          <strong>
            bold
          </strong>
           to show how 
          <em>
            markdown
          </em>
           is stored.
        </div>
      </div>
    </div>
  </div>
</intersect-stub>
`;

exports[`thread-message-feed chat-item snapshot message flags loading 1`] = `
<intersect-stub>
  <div
    class="chat-item burstStart"
    role="listitem"
  >
    <div
      class="chat-item__container"
    >
      <div
        class="chat-item__aside"
      >
        <div
          class="chat-item__avatar"
        >
          <avatar-stub
            size="s"
            user="[object Object]"
          />
        </div>
      </div>
       
      <chat-item-actions-stub
        message="[object Object]"
      />
       
      <div
        class="chat-item__content"
      >
        <div
          class="chat-item__details"
        >
          <div
            class="chat-item__from"
          >
            Tomas Vik
          </div>
           
          <div
            class="chat-item__username"
          >
            @viktomas_gitlab
          </div>
           
          <a
            class="chat-item__time"
            href="/abc/def?at=5d147ea84dad9dfbc522317a"
            title="May 18, 2016 3:48 AM"
          >
            May 18 2016 03:48
          </a>
           
          <loading-spinner-stub
            class="message-loading-icon"
          />
        </div>
         
        <div
          class="chat-item__text"
        >
          <span
            class="mention"
            data-link-type="mention"
            data-screen-name="MadLittleMods"
          >
            @MadLittleMods
          </span>
            Example message using a bit of  
          <code>
            code
          </code>
           and 
          <strong>
            bold
          </strong>
           to show how 
          <em>
            markdown
          </em>
           is stored.
        </div>
      </div>
    </div>
  </div>
</intersect-stub>
`;

exports[`thread-message-feed chat-item snapshot message flags unread 1`] = `
<intersect-stub>
  <div
    class="chat-item burstStart unread"
    role="listitem"
  >
    <div
      class="chat-item__container"
    >
      <div
        class="chat-item__aside"
      >
        <div
          class="chat-item__avatar"
        >
          <avatar-stub
            size="s"
            user="[object Object]"
          />
        </div>
      </div>
       
      <chat-item-actions-stub
        message="[object Object]"
      />
       
      <div
        class="chat-item__content"
      >
        <div
          class="chat-item__details"
        >
          <div
            class="chat-item__from"
          >
            Tomas Vik
          </div>
           
          <div
            class="chat-item__username"
          >
            @viktomas_gitlab
          </div>
           
          <a
            class="chat-item__time"
            href="/abc/def?at=5d147ea84dad9dfbc522317a"
            title="May 18, 2016 3:48 AM"
          >
            May 18 2016 03:48
          </a>
           
          <!---->
        </div>
         
        <div
          class="chat-item__text"
        >
          <span
            class="mention"
            data-link-type="mention"
            data-screen-name="MadLittleMods"
          >
            @MadLittleMods
          </span>
            Example message using a bit of  
          <code>
            code
          </code>
           and 
          <strong>
            bold
          </strong>
           to show how 
          <em>
            markdown
          </em>
           is stored.
        </div>
      </div>
    </div>
  </div>
</intersect-stub>
`;

exports[`thread-message-feed chat-item snapshot not logged in user (no message actions) 1`] = `
<intersect-stub>
  <div
    class="chat-item burstStart deleted"
    role="listitem"
  >
    <div
      class="chat-item__container"
    >
      <div
        class="chat-item__aside"
      >
        <div
          class="chat-item__avatar"
        >
          <avatar-stub
            size="s"
            user="[object Object]"
          />
        </div>
      </div>
       
      <!---->
       
      <div
        class="chat-item__content"
      >
        <div
          class="chat-item__details"
        >
          <div
            class="chat-item__from"
          >
            Tomas Vik
          </div>
           
          <div
            class="chat-item__username"
          >
            @viktomas_gitlab
          </div>
           
          <a
            class="chat-item__time"
            href="/abc/def?at=5d147ea84dad9dfbc522317a"
            title="May 18, 2016 3:48 AM"
          >
            May 18 2016 03:48
          </a>
           
          <!---->
        </div>
         
        <div
          class="chat-item__text"
        >
          
          
          
          <i>
            This message was deleted
          </i>
        </div>
      </div>
    </div>
  </div>
</intersect-stub>
`;

exports[`thread-message-feed chat-item snapshot with default props 1`] = `
<intersect-stub>
  <div
    class="chat-item burstStart"
    role="listitem"
  >
    <div
      class="chat-item__container"
    >
      <div
        class="chat-item__aside"
      >
        <div
          class="chat-item__avatar"
        >
          <avatar-stub
            size="s"
            user="[object Object]"
          />
        </div>
      </div>
       
      <chat-item-actions-stub
        message="[object Object]"
      />
       
      <div
        class="chat-item__content"
      >
        <div
          class="chat-item__details"
        >
          <div
            class="chat-item__from"
          >
            Tomas Vik
          </div>
           
          <div
            class="chat-item__username"
          >
            @viktomas_gitlab
          </div>
           
          <a
            class="chat-item__time"
            href="/abc/def?at=5d147ea84dad9dfbc522317a"
            title="May 18, 2016 3:48 AM"
          >
            May 18 2016 03:48
          </a>
           
          <!---->
        </div>
         
        <div
          class="chat-item__text"
        >
          <span
            class="mention"
            data-link-type="mention"
            data-screen-name="MadLittleMods"
          >
            @MadLittleMods
          </span>
            Example message using a bit of  
          <code>
            code
          </code>
           and 
          <strong>
            bold
          </strong>
           to show how 
          <em>
            markdown
          </em>
           is stored.
        </div>
      </div>
    </div>
  </div>
</intersect-stub>
`;
